File: common/src/main/kotlin/cn/jawbreakers/candycraftce/block/SeaweedBlock.java
package cn.jawbreakers.candycraftce.block;

import com.valentin4311.candycraftmod.registry.CCBlocks;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.tags.FluidTags;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.SimpleWaterloggedBlock;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;

public class SeaweedBlock extends LegacyMetadataBlock.Plant implements SimpleWaterloggedBlock {
	private final boolean canStack;

	public SeaweedBlock(boolean canStack, Properties properties) {
		super(properties);
		this.canStack = canStack;
	}

	@Override
	public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
		BlockState below = level.getBlockState(pos.below());
		BlockState above = level.getBlockState(pos.above());
		boolean bottomValid = below.is(CCBlocks.FLOUR.get())
				|| below.is(CCBlocks.PUDDING.get())
				|| below.is(CCBlocks.CANDY_FARMLAND.get())
				|| below.is(CCBlocks.SUGAR_SAND.get())
				|| canStack && below.is(this);
		boolean topValid = above.getFluidState().is(FluidTags.WATER) || canStack && above.is(this);
		return bottomValid && topValid && level.getFluidState(pos).is(FluidTags.WATER) && hasOpaqueOrWaterNeighbors(level, pos);
	}

	private static boolean hasOpaqueOrWaterNeighbors(LevelReader level, BlockPos pos) {
		for (Direction direction : Direction.Plane.HORIZONTAL) {
			BlockPos neighborPos = pos.relative(direction);
			BlockState neighborState = level.getBlockState(neighborPos);
			if (!neighborState.getFluidState().is(FluidTags.WATER)
					&& !neighborState.isFaceSturdy(level, neighborPos, direction.getOpposite())) {
				return false;
			}
		}
		return true;
	}

	@Override
	public BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor level, BlockPos pos, BlockPos neighborPos) {
		level.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(level));
		if (!state.canSurvive(level, pos)) {
			return Blocks.WATER.defaultBlockState();
		}
		return super.updateShape(state, direction, neighborState, level, pos, neighborPos);
	}

	@Override
	public FluidState getFluidState(BlockState state) {
		return Fluids.WATER.getSource(false);
	}
}

File: common/src/main/kotlin/cn/jawbreakers/candycraftce/block/SeaweedBlock.kt
package cn.jawbreakers.candycraftce.block

import cn.jawbreakers.candycraftce.registry.CBlockTags
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.tags.FluidTags
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.LevelAccessor
import net.minecraft.world.level.LevelReader
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.SimpleWaterloggedBlock
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.material.FluidState
import net.minecraft.world.level.material.Fluids

class SeaweedBlock(
    private val canStack: Boolean,
    properties: Properties,
) : CandyPlantBlock(properties), SimpleWaterloggedBlock {
    override fun mayPlaceOn(state: BlockState, level: BlockGetter, pos: BlockPos): Boolean {
        return state.`is`(CBlockTags.seaweed_soil) || (canStack && state.`is`(this))
    }

    override fun canSurvive(state: BlockState, level: LevelReader, pos: BlockPos): Boolean {
        if (mayPlaceOn(level.getBlockState(pos.below()), level, pos.below())) {
            if (state.fluidState.`is`(FluidTags.WATER)) {
                val above = level.getBlockState(pos.above())
                if (above.fluidState.`is`(FluidTags.WATER) || canStack && above.`is`(this)) {
                    return true
                }
            }
        }
        return false
    }

    override fun updateShape(
        state: BlockState,
        direction: Direction,
        neighborState: BlockState,
        level: LevelAccessor,
        pos: BlockPos,
        neighborPos: BlockPos,
    ): BlockState {
        level.scheduleTick(pos, Fluids.WATER, Fluids.WATER.getTickDelay(level))
        if (!state.canSurvive(level, pos)) {
            return Blocks.WATER.defaultBlockState()
        }
        return super.updateShape(state, direction, neighborState, level, pos, neighborPos)
    }

    @Deprecated("Deprecated in Java")
    override fun getFluidState(state: BlockState): FluidState {
        return Fluids.WATER.getSource(false)
    }
}

